File: algo/69. Sqrt(x).js
// @9 Binary Search

/**
 * @param {number} x
 * @return {number}
 */
var mySqrt = function(x) {
    var start = 0;
    var end = x;
    while (start+1<end) {
        var mid = Math.floor((end-start)/2) + start;
        if (mid*mid == x) {
            return mid;
        } else if (mid*mid < x) {
            start = mid;
        } else {
            end = mid;
        }
    }
    
    if (end**end <= x) {
        return end;
    } else {
        return start;
    }
};

File: round2/69. Sqrt(x).cs
public class Solution {
    public int MySqrt(int x) {
        if (x == 0) return 0;
        int start = 0, end = x;
        while (start+1<end) {
            int mid = (end-start)/2 + start;
            if (x == mid * mid) return mid;
            else if ((double) (x / mid) / (double) mid > 1) start = mid;
            else end = mid;
            //Console.WriteLine(mid);
        }
        if ((double) (x / end) / (double) end >= 1) return end;
        else return start;
    }
}